src/Spanner/Database.php

```diff
@@ -0,0 +1,1313 @@
+     *         $t->commit();
+     *     } else {
+     *         $t->rollback();
+     *     }
+     * });
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @see https://cloud.google.com/spanner/reference/rpc/google.spanner.v1#google.spanner.v1.BeginTransactionRequest BeginTransactionRequest
+     * @see https://cloud.google.com/spanner/docs/transactions Transactions
+     * @codingStandardsIgnoreEnd
+     *
+     * @param callable $operation The operations to run in the transaction.
+     *        **Signature:** `function (Transaction $transaction)`.
+     * @param array $options [optional] {
+     *     Configuration Options
+     *
+     *     @type int $maxRetries The number of times to attempt to apply the
+     *           operation before failing. **Defaults to ** `3`.
+     *     @type Transaction $transaction If provided, the transaction will be
+     *           passed to the callable instead of attempting to begin a new
+     *           transaction.
+     * }
+     * @return mixed The return value of `$operation`.
+     */
+    public function runTransaction(callable $operation, array $options = [])
+    {
+        $options += [
+            'maxRetries' => self::MAX_RETRIES,
+            'transaction' => null
+        ];
+
+        // There isn't anything configurable here.
+        $options['transactionOptions'] = $this->configureTransactionOptions();
+
+        $session = $this->selectSession(SessionPoolInterface::CONTEXT_READWRITE);
+
+        $attempt = 0;
+        $startTransactionFn = function ($session, $options) use (&$attempt) {
+            if ($attempt === 0 && $options['transaction'] instanceof Transaction) {
+                $transaction = $options['transaction'];
+            } elseif ($attempt === 0 && $options['transaction']) {
+                throw new \InvalidArgumentException('Given transaction must be an instance of Transaction.');
+            } else {
+                $transaction = $this->operation->transaction($session, $options);
+            }
+
+            $attempt++;
+            return $transaction;
+        };
+
+        $delayFn = function (\Exception $e) {
+            if (!($e instanceof AbortedException)) {
+                throw $e;
+            }
+
+            $delay = $e->getRetryDelay();
+            time_nanosleep($delay['seconds'], $delay['nanos']);
+        };
+
+        $commitFn = function ($operation, $session, $options) use ($startTransactionFn) {
+            $transaction = call_user_func_array($startTransactionFn, [
+                $session,
+                $options
+            ]);
+
+            return call_user_func($operation, $transaction);
+        };
+
+        $retry = new Retry($options['maxRetries'], $delayFn);
+
+        try {
+            return $retry->execute($commitFn, [$operation, $session, $options]);
+        } finally {
+            $session->setExpiration();
+        }
+    }
+
+    /**
+     * Create and return a new read/write Transaction.
```

Comment:
We don't wish to limit ways in which users can manage their application. We've imagined several scenarios in which users could prefer to manage a transaction manually such as if an application had existing retry logic which the user wished to leverage.